### PR TITLE
Fix "Only X left" functionality

### PR DESCRIPTION
**Pull request by aleron75**
This PR is intended to fix issue #1165 

- Add `InventoryFrontendUi` module to implement the fix
- Pluginize methods used by `\Magento\CatalogInventory\Block\Stockqty\AbstractStockqty` instances
- Update `InventoryConfiguration` to access the stock threshold qty 

This PR is also intended to fix issue #1349

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=1348) <br/>All committers have signed the CLA.

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: please fix the PHP Doc Block

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: please provide doc block for constructor

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: please specify parameter for this method in doc block

File: app/code/Magento/InventoryFrontendUi/composer.json
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/composer.json]: it's better to rename module to InventorySalesFrontendUi

File: app/code/Magento/InventoryFrontendUi/composer.json
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/composer.json]: please add the reference to this module in "require" section of the root composer.json

File: app/code/Magento/InventoryFrontendUi/etc/module.xml
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/etc/module.xml]: no need to use `setup_version` 

that's true for other Inventory modules as well

File: app/code/Magento/InventoryFrontendUi/etc/di.xml
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/etc/di.xml]: looks like it's copy-paste of previous string

File: app/code/Magento/InventoryFrontendUi/composer.json
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/composer.json]: dependency should be fixed on `module-inventory-configuration-api`

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: static tests will break here

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: there is no declare strict type here

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: protected methods are forbidden in newly created code, just private is allowed to use

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: provide correct doc type

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: let's add `$stockItemConfig->getMinQty()` into the `getProductSalableQty` service

so that we can substitute getStockQtyLeft
with getProductSalableQuantity

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: let's use clean check for Backorder
`getBackorders` instead of `$this->getProductSalableQty->execute($sku, $stockId) > 0`
because even if we have Salable Qty = 5 , and getStockThresholdQty = 10
that's not correct to say that only 5 left, because having backorder enabled we can sell infinite number

**Review by maghamed — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (aleron75) [reply app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: I'm not sure I understood your proposal, I changed the code like this:
```
return $stockItemConfig->getBackorders() === StockItemConfigurationInterface::BACKORDERS_NO
            && $this->getProductSalableQty->execute($sku, $stockId) > 0
            && $this->getStockQtyLeft($sku, $stockId) <= $stockItemConfig->getStockThresholdQty();
```
If backorders are enabled, the first condition will fail and the message won't be dispayed.

If backorders are disabled, the first condition will return `true` and the other two conditions check that the product is salable and that the salable qty is lett or equal than the threshold.

Is that correct?

File: app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (aleron75) [reply app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: Since incorporating the min qty subtraction into the `getProductSalableQty` service requires refactoring of different pieces of code, I opened the issue #1349, assigned it to myself, not to block this PR.

File: app/code/Magento/InventorySalesApi/Api/GetProductSalableQtyInterface.php
Comment (maghamed) [inline app/code/Magento/InventorySalesApi/Api/GetProductSalableQtyInterface.php]: also need to add InputExcepton here 
as implementation may throw 
```
            throw new InputException(
                __('Can\'t check requested quantity for products without Source Items support.')
            );
```

File: app/code/Magento/InventorySalesFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php
Comment (maghamed) [inline app/code/Magento/InventorySalesFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php]: we don't need this check 
` && $this->getProductSalableQty->execute($sku, $stockId) > 0`

because in this case product would not be salable at all

**Review by maghamed — CHANGES_REQUESTED** [review]

**Comment by smoskaluk** [comment]
@aleron75 @maghamed I could not verify this PR because message 'Only x left' appear on product page only after flushing the cache. Please re-check

**Review by maghamed — APPROVED** [review]

**Pull request merged**
